### PR TITLE
SET_INDESTRUCTIBLE opcode

### PR DESCRIPTION
**Pull request by vbuterin**
Add a `SET_INDESTRUCTIBLE` opcode that prevents the contract from calling `SELFDESTRUCT`.

When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/eip-2937.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+discussions-to: https://ethereum-magicians.org/t/eip-2937-set-indestructible/4571
+status: Draft
+type: Standards Track
+category: Core
+created: 2020-09-04
+---
+
+## Simple Summary
+
+Add a `SET_INDESTRUCTIBLE (0xA8)` opcode that prevents the contract from calling `SELFDESTRUCT (0xff)`.
```
Comment (lightclient) [inline EIPS/eip-2937.md:14]:
```suggestion
Add a `SET_INDESTRUCTIBLE (0xA8)` opcode that prevents the contract from calling `SELFDESTRUCT (0xFF)`.
```

File: EIPS/eip-2937.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+## Specification
+
+Add a context-wide variable `locals.indestructible` (ie. a variable that lives in the same "space" as memory, stack, PC...), initialized to `False`.
```
Comment (lightclient) [inline EIPS/eip-2937.md:18]:
```suggestion
Add a context-wide variable `locals.indestructible` (i.e. a variable that lives in the same "space" as memory, stack, PC...), initialized to `False`.
```

File: EIPS/eip-2937.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+Add a `SET_INDESTRUCTIBLE` opcode at `0xA8`, with gas cost `G_base`, that sets `locals.indestructible` to `True`. If in the current execution context `locals.indestructible` is `True`, the `SELFDESTRUCT` opcode throws an exception.
+
+## Motivation
```
Comment (lightclient) [inline EIPS/eip-2937.md:22]: Motivation is usually above specification.

**Review by lightclient — APPROVED** [review]

File: EIPS/eip-2937.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+type: Standards Track
+category: Core
+created: 2020-09-04
+---
+
+## Simple Summary
+
+Add a `SET_INDESTRUCTIBLE (0xA8)` opcode that prevents the contract from calling `SELFDESTRUCT (0xff)`.
+
+## Specification
```
Comment (MicahZoltu) [inline EIPS/eip-2937.md:16]:
```suggestion
## Abstract

## Specification
```
Try to match the template whenever possible as it allows us to do fancy things with automated tooling in the future.

File: EIPS/eip-2937.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+Add a `SET_INDESTRUCTIBLE` opcode at `0xA8`, with gas cost `G_base`, that sets `locals.indestructible` to `True`. If in the current execution context `locals.indestructible` is `True`, the `SELFDESTRUCT` opcode throws an exception.
+
+## Motivation
+
+The intended use case would be for contracts to make their first byte of code be the `SET_INDESTRUCTIBLE` opcode if they wish to serve as libraries that guarantee to users that their code will exist unmodified forever. This is useful in account abstraction as well as other contexts.
+
+Unlike EIPs that disable the `SELFDESTRUCT` opcode entirely, this EIP does not modify behavior of any existing contracts.
+
+## Forwards Compatibility and Security
```
Comment (MicahZoltu) [inline EIPS/eip-2937.md:28]:
```suggestion
## Rationale

TBD

## Backwards Compatibility

TBD

## Security Considerations

TBD
```

File: EIPS/eip-2937.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+This breaks forward compatibility with _some_ forms of state rent, which would simply delete contracts that get too old without paying some maintenance fee. However, this is not the case with all state size control schemes; for example this is not an issue if we use [ReGenesis](https://ledgerwatch.github.io/regenesis_plan.html).
+
+If `SELFDESTRUCT` is ever removed in the future, this EIP would simply become a no-op.
```
Comment (MicahZoltu) [inline EIPS/eip-2937.md:32]: I'm not personally a fan of forward compatibility stuff in an EIP.  This feels like something much more appropriate in the discussion-to link, or in the HF inclusion discussions.  If you imagine reading this EIP 5 years in the future after it has been implemented, this section won't really provide any significant value to those readers.

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
Formatting stuff needs to be addressed before merge.

**Review by MicahZoltu — APPROVED** [review]

**Pull request merged**